### PR TITLE
gtkclipblock: init at 0.2.2

### LINKED ISSUE #271603: Package request: gtkclipblock
**Project description**
>  A hack to prevent GTK programs from interacting with the primary clipboard (aka "primary selection")

Most useful if this can be added to `firefox`'s ld preload, so wayland users on kde who disabled paste with middle click are able to copy links from the search bar.

**Metadata**

* homepage URL: https://github.com/notpeelz/gtkclipblock
* source URL: https://github.com/notpeelz/gtkclipblock
* license: lgpl3
* platforms: unix

**Priorities**

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by UARTman**
## Description of changes
Adds [gtkclipblock](https://github.com/notpeelz/gtkclipblock). 

It is a hack to prevent GTK programs from interacting with the primary clipboard (aka "primary selection"), written to resolve a currently open [Firefox bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1791417) and requested in and closes https://github.com/NixOS/nixpkgs/issues/271603. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (eclairevoyant) [inline pkgs/tools/misc/gtkclipblock/default.nix:17]:
```suggestion
    hash = "sha256-2ftEV3TMS3HT7f96k+Pwt3Mm31fVEXcHpcbbz05jycU=";
```

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (eclairevoyant) [inline pkgs/tools/misc/gtkclipblock/default.nix:23]:
```suggestion
    hash = "sha256-ok/D7M0KekN8zf8AzhcOLtedbYVRHHv3m9zEHsJfcPM=";
```

File: pkgs/top-level/all-packages.nix
Comment (eclairevoyant) [inline pkgs/top-level/all-packages.nix:8977]: any reason to use gcc13 specifically?

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix
Comment (UARTman) [reply pkgs/top-level/all-packages.nix:8977]: It doesn't build with `stdenv` and `gcc12Stdenv`, so my hypothesis is that it requires some feature introduced in GCC 13 (though I haven't dug too deep into this).

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (UARTman) [reply pkgs/tools/misc/gtkclipblock/default.nix]: I've applied the edit in both places.

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/gtkclipblock/default.nix:31]: Why the are you adding cmake AND meson?

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/gtkclipblock/default.nix:12]: Are you afraid of blank lines?

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/gtkclipblock/default.nix:35]: A subproject that gits a subproject that gits a subproject.

How can such a chain work?

Comment (UARTman) [reply pkgs/tools/misc/gtkclipblock/default.nix:35]: `gtkclipblock` is a meson project with a git submodule to `funchook-helper` in its `subprojects` folder (`subprojects/funchook-helper`). `funchook-helper` is another meson project (hosted in a different repo) with a git submodule to `funchook` in *its* `subprojects` folder (`subprojects/funchook`). `funchook` is a CMake project that uses FetchContent to get `distorm` (the fourth git repo involved), hence the patch that redirects it to a fetchFromGitHub (because FetchContent can't fetch it when sandboxed).

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (UARTman) [reply pkgs/tools/misc/gtkclipblock/default.nix:31]: Because `gtkclipblock` submodules and builds `funchook-helper`, which submodules and builds `funchook`, which is a CMake project.

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (UARTman) [reply pkgs/tools/misc/gtkclipblock/default.nix]: Added blank lines.

**Comment by huantianad** [comment]
What's the recommended way of actually using this library with firefox? Is it different than described in the project's readme, or is there a way to do it via Nix entirely?

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (AndersonTorres) [reply pkgs/tools/misc/gtkclipblock/default.nix:34]: This can be confusing, since both Meson and Cmake modify the build and install phases, and I don't know which one will be executed.

**Comment by UARTman** [comment]
> What's the recommended way of actually using this library with firefox? Is it different than described in the project's readme, or is there a way to do it via Nix entirely?

I know for certain that using it as described in the readme (`LD_PRELOAD=/nix/store/990wrh4f1ck0baiv6a0l93pc0wqmsl1a-gtkclipblock-0.2.2/lib/libgtkclipblock.so GTKCLIPBLOCK_HOOK=1 firefox`) does work and fix the problem. The only way I can come up with to do it via nix entirely involves modifying Firefox or creating a wrapper package. I don't know if that is warranted (this is a hack, after all) or in scope (should probably go into separate PR) and also don't quite feel up to messing with firefox yet.

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (UARTman) [reply pkgs/tools/misc/gtkclipblock/default.nix:34]: Is there any way to make it less confusing/pass CMake to meson in a way that won't modify the build/install phases? Or at least get the commands that run each build/install to see what happens?

I am also a bit confused as to how that ambiguity affects us practically speaking, because I definitely know the package builds/installs correctly (using meson) and the artifact produced works as advertised...

Comment (AndersonTorres) [reply pkgs/tools/misc/gtkclipblock/default.nix:34]: I will ask about it on Discourse.

Comment (AndersonTorres) [reply pkgs/tools/misc/gtkclipblock/default.nix:34]: Usually it is done via `dontUseCmakeConfigure`.

Comment (UARTman) [reply pkgs/tools/misc/gtkclipblock/default.nix:34]: Thanks! I've set it.

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:8977]: I suggest to use the new by-name hierarchy.
```suggestion
  gtkclipblock = callPackage ../by-name/gt/gtkclipblock/package.nix { 
   stdenv = gcc13Stdenv; 
  };
```

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/gtkclipblock/default.nix:36]: Just to know.
Is GTK2 an optional dependency?
I ask this because GTK 2 is deprecated.

File: pkgs/top-level/all-packages.nix
Comment (UARTman) [reply pkgs/top-level/all-packages.nix:8977]: Done.

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (AndersonTorres) [reply pkgs/tools/misc/gtkclipblock/default.nix]: Lemme see if I understood.

The options below are mutually exclusive? Only one can be checked at a time?

- [ ] gtk2
- [ ] gtk3
- [ ] gtk4
- [ ] gtk2 and gtk3 and gtk4

Comment (UARTman) [reply pkgs/tools/misc/gtkclipblock/default.nix]: That's the options gtkclipblock gives, yes. The only way to support gtk3 and gtk4 without gtk2 is to edit `meson.build` and comment out a part of it. Which is rather hacky, but can be done if we'd rather not support gtk2 here.

Comment (UARTman) [reply pkgs/tools/misc/gtkclipblock/default.nix]: Here's the derivation that does this, is this alright or is it too hacky? 

https://gist.github.com/UARTman/a39e8f7931019044f5005127ab6bbb7b

Comment (AndersonTorres) [reply pkgs/tools/misc/gtkclipblock/default.nix]: Since it requires editing the meson file to such a huge extent, it is better to suggest this change to the upstream project instead of keeping the hack in Nixpkgs.

For now let's bundle everything.

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (pbsds) [inline pkgs/by-name/gt/gtkclipblock/package.nix:46]:
```suggestion
    description = "A hack to prevent GTK programs from interacting with the primary clipboard. To use it, prepend it to LD_PRELOAD";
```

**Review by pbsds — COMMENTED** [review]
Please squash the commits into a single commit

---

Result of `nixpkgs-review pr 272092` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>gtkclipblock</li>
  </ul>
</details>

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/gt/gtkclipblock/package.nix:46]: We don't - or shouldn't - do hints of usage in description.

Comment (pbsds) [reply pkgs/by-name/gt/gtkclipblock/package.nix:46]: https://search.nixos.org/packages?query=LD_PRELOAD

Comment (pbsds) [reply pkgs/by-name/gt/gtkclipblock/package.nix:46]:
```suggestion
    description = "A LD_PRELOAD hack to prevent GTK programs from interacting with the primary clipboard";
```

Comment (pbsds) [reply pkgs/by-name/gt/gtkclipblock/package.nix:46]: IMO one of the great strengths of nixpkgs is how much domain knowledge it contains and makes discoverable

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (pbsds) [inline pkgs/by-name/gt/gtkclipblock/package.nix:49]:
```suggestion
    platforms = platforms.unix;
    broken = !stdenv.isx86_64;
```

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (pbsds) [reply pkgs/by-name/gt/gtkclipblock/package.nix:46]: I'll let the maintainer, @UARTman, make the call :)

Comment (AndersonTorres) [reply pkgs/by-name/gt/gtkclipblock/package.nix:46]: Your second suggestion is way better.

We should not tell people how to use a software we package. 
Think on the absurdity of something like _"Emacs is a text editor. Use it to contribute to GitLab"_.

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/gt/gtkclipblock/package.nix:49]: Broken or unsupported?

File: pkgs/tools/misc/gtkclipblock/default.nix
Comment (huantianad) [reply pkgs/tools/misc/gtkclipblock/default.nix]: Looking at the patch, can't we just edit the if statement to never execute?

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (pbsds) [reply pkgs/by-name/gt/gtkclipblock/package.nix:49]: If the latter

```suggestion
    platforms = [ "x86_64-linux" ];
```

**Comment by UARTman** [comment]
> Please squash the commits into a single commit

@pbsds Should I just rebase and do a `git push -f`? Or is there a better way? Also, should I leave the maintainers commit or should I just fold everything into one commit?

**Comment by huantianad** [comment]
Rebase and force push is the way to go. Keep the maintainer commit separate 👍

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/gt/gtkclipblock/package.nix:32]:
```suggestion
  inherit pname version src;
```

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/gt/gtkclipblock/package.nix:20]: I believe it conveys better your intention of "factoring the things out":

```suggestion
  pname = "gtkclipblock";
  version = "0.2.2";

  src = fetchFromGitHub {
```

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/gt/gtkclipblock/package.nix:23]:
```suggestion
    rev = "v${version}";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by AndersonTorres — APPROVED** [review]
Squash the commits. Besides, lgtm.

**Review by huantianad — APPROVED** [review]
Looks good to me, would love a firefox option to use this once merged!

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:8977]: With the next staging-next merge, this should become the default.

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/gt/gtkclipblock/package.nix:36]: Is a new dependency on gtk2 really required?

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/gt/gtkclipblock/package.nix:34]: We likely only need 3 or 2 of those. The warnings by meson can be ignored.

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/gt/gtkclipblock/package.nix:32]: We could save the inherit by just inlining things.

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/gt/gtkclipblock/package.nix:36]: @SuperSandro2000 the upstream code is written in an all-or-nothing fashion. If some of these libs are not present the code ignores all of them. 

(Being more accurate, it does not work unless exactly one or three of them are present.)

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/gt/gtkclipblock/package.nix:34]: Cmake is needed as part of building some subprojects.

File: pkgs/by-name/gt/gtkclipblock/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/gt/gtkclipblock/package.nix:32]: This is easier to read and not get distracted.

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 272092` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>gtkclipblock</li>
  </ul>
</details>

LGTM

**Pull request merged**
